lib/matplotlib/layout_engine.py

```diff
@@ -0,0 +1,223 @@
+    """
+    def __init__(self, **kwargs):
+        self._figure = None
+        self.colorbar_gridspec = True
+        self._params = {}
+        self.adjust_compatible = False
+
+    def set(self, **kwargs):
+        raise NotImplementedError
+
+    def get(self):
+        return self._params
+
+    def execute(self):
+        raise NotImplementedError
+
+    def set_figure(self, figure):
+        """
+        Set the figure for the layout manager.
+        """
+        if self._figure == figure:
+            return
+        if self._figure is None:
+            self._figure = figure
+        else:
+            raise ValueError(
+                'Layout managers cannot be attached to more than one figure')
+
+    def get_figure(self):
+        return self._figure
+
+
+class TightLayoutEngine(LayoutEngine):
+    """
+    Implements the ``tight_layout`` geometry management.
+    """
+
+    def __init__(self, *, pad=1.08, h_pad=None, w_pad=None,
+                 rect=(0, 0, 1, 1)):
+        """
+        Initialize tight_layout engine.
+
+        Parameters
+        ----------
+        pad : float, 1.08
+            Padding between the figure edge and the edges of subplots, as a
+            fraction of the font size.
+        h_pad, w_pad : float
+            Padding (height/width) between edges of adjacent subplots.
+            Defaults to *pad*.
+        rect : tuple[float, float, float, float], optional
+            (left, bottom, right, top) rectangle in normalized figure
+            coordinates that the subplots (including labels)
+            will fit into. Defaults to using the entire figure.
+        """
+        super().__init__()
+        self.colorbar_gridspec = True
+        self.adjust_compatible = True
+        self.set(pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
+
+    def execute(self):
+        """
+        Execute tight_layout.
+
+        This decides the subplot parameters given the padding that
+        will allow the axes labels to not be covered by other labels
+        and axes.
+
+        See also: `.figure.Figure.tight_layout` and `.pyplot.tight_layout`.
+        """
+        fig = self._figure
+        info = self._params
+        subplotspec_list = get_subplotspec_list(fig.axes)
+        if None in subplotspec_list:
+            _api.warn_external("This figure includes Axes that are not "
+                               "compatible with tight_layout, so results "
+                               "might be incorrect.")
+        renderer = _get_renderer(fig)
+        with getattr(renderer, "_draw_disabled", nullcontext)():
+            kwargs = get_tight_layout_figure(
+                fig, fig.axes, subplotspec_list, renderer,
+                pad=info['pad'], h_pad=info['h_pad'], w_pad=info['w_pad'],
+                rect=info['rect'])
+        if kwargs:
+            fig.subplots_adjust(**kwargs)
+
+    def set(self, pad=1.08, w_pad=None, h_pad=None, rect=(0, 0, 1, 1)):
```

Comment:
Here (unlike in `__init__`), everything needs to default to None, so that setting e.g. `pad` does not reset rect.

Reply:
Also make these keyword-only, and possibly use kwdefaults below as well.

Reply:
> Here (unlike in `__init__`), everything needs to default to None, so that setting e.g. `pad` does not reset rect.

OK, that is a question.  Should `set` only affect the passed in kwargs, or should it reset any that are not passed?

Reply:
I think matplotlib's standard approach would be to just update what's passed in; if you want to reset everything you may just as well instantiate a new engine, right?

Reply:
Fair - done...